next: Add AST node for For loops

dlongnecke-cray opened the pull request:
next: Add AST node for For loops (#17658)

This is for `compiler/next`.

Add AST interfaces for for loops, as well as parser rules for parsing
for loops that appear at a statement level.

Add a new abstract class `IndexableLoop` that is a child of `Loop`.
It represents any kind of loop which is "indexable", such as `For`,
`Forall`, `Coforall`. It contains methods to get the index variable
and iterand of the loop.

After discussion, we settled on having one index variable and one
iterand per IndexableLoop (as opposed to having N iterands and N
index variables). The `zip()` construct will have its own AST node
which will represent it as an iterand.

Reviewed by @mppf and @vasslitvinov. Thanks!

TESTING:

- [x] All `compiler/next` tests pass

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

mppf commented on compiler/next/include/chpl/uast/Loop.h:
```diff
@@ -0,0 +1,84 @@
+    return true;
+  }
+  void expressionMarkUniqueStringsInner(Context* context) const {}
+  int32_t statementChildNum_;
```
Can we name this `loopBodyChildNum` or something like that?

mppf commented on compiler/next/lib/uast/For.cpp:
```diff
@@ -0,0 +1,139 @@
+    assert(child->isExpression());
+  }
+#endif
+}
```
Let's move the constructor to the header (see above comment). Also, replace the loop with `    assert(isExpressionASTList(children_));`.

mppf commented on compiler/next/lib/uast/For.cpp:
```diff
@@ -0,0 +1,139 @@
+}
+
+void For::markUniqueStringsInner(Context* context) const {
+  expressionMarkUniqueStringsInner(context);
```
This should call loopMarkUniqueStringsInner

mppf commented on compiler/next/lib/uast/For.cpp:
```diff
@@ -0,0 +1,139 @@
+                      bool usesDo,
+                      bool expressionLevel,
+                      bool param) {
+#ifndef NDEBUG
```
`assert` already is a macro that checks this. So you can remove the ifndef here since the contents are only assert calls.

mppf commented on compiler/next/lib/uast/For.cpp:
```diff
@@ -0,0 +1,139 @@
+#ifndef NDEBUG
+  assert(indexVar.get() != nullptr);
+  assert(iterand.get() != nullptr);
+  assert(param ? !expressionLevel : true);
```
I'm not used to seeing the ternary operator in an assert and it would be faster for me to understand if you write it as `assert(param==false || expressionLevel==false)`. Alternatively I think it would be even easier to understand as `if (param) assert(!expressionLevel);`.

mppf commented on compiler/next/lib/uast/For.cpp:
```diff
@@ -0,0 +1,139 @@
+  return toOwned(ret);
+}
+
+owned<For> For::build(Builder* builder,
```
Oh, I see. This version does not have an index var. Let's just make it call the above and pass `nullptr` as the index variable (adjusting the asserts appropriately).

mppf commented on compiler/next/lib/uast/Loop.cpp:
```diff
@@ -0,0 +1,44 @@
+  assert(statementChildNum < this->numChildren());
+#endif
+
+}
```
Move the constructor to the header

mppf commented on compiler/next/include/chpl/uast/For.h:
```diff
@@ -0,0 +1,115 @@
+  /**
+    Returns true if this for loop is param.
+  */
+  bool param() const {
```
In this case I think it'd be nicer to name this function `isParamLoop` or `isParam`. `param` makes it sound to me like it will return "the param" which doesn't really make sense.

mppf commented on compiler/next/include/chpl/uast/For.h:
```diff
@@ -0,0 +1,115 @@
+  /**
+    Returns true if this for loop appears at the expression level. 
+  */
+  bool expressionLevel() const {
```
I think this one would be better named `isExpressionLevel` or `isLoopExpression`. The problem with `expressionLevel` is that it sounds like it should return an `int` indicating the number of the level. (But that doesn't make sense).

mppf commented on compiler/next/include/chpl/uast/For.h:
```diff
@@ -0,0 +1,115 @@
+  /**
+    Returns the index variable of this for loop.
+  */
+  const Expression* indexVar() const {
```
It looks like you are currently representing the `indexVar` as an `Identifier` but it's actually a new variable, so I would expect that the loop contain a `VariableDecl` for it.

Then, this function should probably have two variants:
 * `indexVariableDecl` returns a `VariableDecl*`
 * `indexVariable` returns a `Variable*`
 
I don't currently know of a reason why the VariableDecl would be needed but it might be and it's easy enough to include. It will not be easy to go from the indexVariable back to the Decl (since uast does not have links to the parent).

mppf replied:
I had a side conversation with @lydia-duncan about this and the conclusion of that conversation is:
 * the normal case is wanting to get to the `Sym` for the declared variable (so `indexVariable` should return `Variable*`)
 * we would like to be able to access the `Decl` in case it is needed, so would want the 2nd variant (e.g. `indexVariableDecl` above). This is important because the uAST nodes don't have links back to parent nodes.

mppf commented on compiler/next/include/chpl/uast/Loop.h:
```diff
@@ -0,0 +1,84 @@
+  virtual ~Loop() override = 0; // this is an abstract base class
+
+  /**
+    Return a way to iterate over the statements of this loop.
```
See https://github.com/Cray/chapel-private/issues/1978

dlongnecke-cray commented:
Hi @mppf would you do a quick scan over things to make sure I've gotten technicals right?

To @vasslitvinov as the resident loop master, would you mind reviewing this?

Some things to note:

- I've added `IndexableLoop` as a subclass of `Loop`, its children are things like `Forall`, `Coforall`, `For`
- Desugaring of tuple index variables (e.g. `for (a, b) in foo do ...` is implicit in the number of index variables
- You can fetch the `i'th` index variable of an `IndexableLoop` using `indexVariable(int i)`
- I've added a new variable storage kind called `INDEX`, which represents an index variable with an indeterminate storage kind

vasslitvinov commented:
> I've added `IndexableLoop` as a subclass of Loop, its children are things like Forall, Coforall, For

This uses inheritance to reflect one of the factorable-out properties of loops. Which will prevent us to use inheritance for other such properties, for example with-clauses -- which, for example, are available for Cobegins, which are not Indexable.

Please consider using a has-a relationship instead of is-a. Specifically, I propose to create a uAST node, say `LoopIndexing`, that will represent the indexing syntax and that all indexable loops will contain as a child. Similarly, a `WithClause` would factor out the with-syntax and be an optional child of Forall et al.

vasslitvinov commented on compiler/next/include/chpl/uast/For.h:
```diff
@@ -0,0 +1,110 @@
+  .. code-block:: chapel
+
+      // Example 1:
+      for i in foo {
```
I'd expect `foo` to have parens here.

dlongnecke-cray replied:
What about something that has a `these`?

vasslitvinov replied:
Of course. To me `foo` is more a function than a variable. If we prefer going without parens, I suggest using something like `myRange` instead of `foo`.

dlongnecke-cray replied:
Sounds good, thanks!

vasslitvinov commented on compiler/next/include/chpl/uast/IndexableLoop.h:
```diff
@@ -0,0 +1,123 @@
+
+  int8_t iterandChildNum_;
+  int8_t indexVariableDeclStartChildNum_;
+  int numIndexVariables_;
```
This is good for the initial prototype.

In the long run, we want to represent complex indexing expressions, ex. `for (a, (b,c), _) in ....`

So it may be easier to have a single Expression to represent index variable(s). This expression will be either Identifier or Tuple.

dlongnecke-cray replied:
You're right that this representation does not work for something like `(a, (b,c), _)`, because all nesting is lost. @mppf wanted me to introduce variable declarations for each loop index. Personally I think it would just be easier to leave the expression as the index variable and just interpret the expression appropriately when resolving instead of introducing decls.

dlongnecke-cray replied:
Would it suffice to just store the original index expression in addition to the index variable declarations? It could be accessed via `indexExpression` and it could be used to type the index variables during resolution (In the `(a, (b, c), _)` case). Perhaps that would be a good compromise.

mppf replied:
We have tuple variable declarations too; you can write `var (a,b) = (1,2);`. So, I think we need to have some sort of TupleDeclaration uAST node that doesn't exist yet and we should use that here as well. So, I stand by my request that the ForLoop include some sort of Decls for the index variables; the case `for (a, (b,c), _) in ...` will just use `Decls` that we haven't defined yet.

I think it's important that in the common cases along the lines of `for (a,b) in zip(A, B)` that there are no tuples introduced into the representation. Of course there are other cases where we need to handle tuples - but in those cases the compiler is more or less forced into generating code that does things like tuple packing or unpacking. It's much easier to add the tuple representation in later during compilation than to remove it.

vasslitvinov commented on compiler/next/include/chpl/uast/IndexableLoop.h:
```diff
@@ -0,0 +1,123 @@
+  const Expression* iterand() const {
+    if (iterandChildNum_ < 0) return nullptr;
+    return (const Expression*)child(iterandChildNum_);
+  }
```
Again, OK for now, however we will need to support zippering as well. Given that zippering does not nest, it makes sense to have each zippered iterand be a direct child of Indexable. (I am contrasting index variables here.)

Note that (loop is zippered) is not necessarily the same as (#iterables > 1), so need a dedicated boolean for that.

dlongnecke-cray replied:
Can't you just figure out if the loop is zippered by looking at the iterand?

mppf replied:
I agree with @vasslitvinov here that we should have an interface here to get the i'th zippered iterand etc. Zippering is syntactical so should be reflected here.

Of course we could do something like introduce a `Zip` uAST node but since it only comes up in loops and there can only be one per loop I think it makes more sense to have `iterand(int i)` and `numIterands()` etc.

dlongnecke-cray replied:
This whole time I assumed that there would be a `Zip` AST node. I think it appears in often enough places that we could just make a `Zip` as a subclass of `Call`.

vasslitvinov reviewed (approved):
See my comments.

mppf commented:
> This uses inheritance to reflect one of the factorable-out properties of loops. Which will prevent us to use inheritance for other such properties, for example with-clauses -- which, for example, are available for Cobegins, which are not Indexable.

I'm not necessarily against such a has-a relationship instead of is-a; but in C++ it is possible to use multiple inheritance. Perhaps there is some reason we would rule that out but I don't know of it yet.

Additionally I tend to think of the task intent support (WithClause in your example) as something that makes sense in a "has-a" relationship because we necessarily need to handle multiple task/forall intents. So the uAST node supporting these basically just needs to list them (just as a Function has Formals). Vs. being an "indexable loop" seems to me to be more fundamental to the nature of the loop and affects nearly every aspect of the uAST node's interface. (E.g.; a `while` loop has a condition and a body; but an "indexable loop" has to handle index variables, iterable expressions, and zippering).

So, all that said, my recommendation is to keep IndexableLoop as a base class but to make "has-with-clauses" something that is mostly handled by a "has-a" relationship (we can model these with some WithClause uAST nodes). In that world, we can consider if IndexableLoop also includes methods to access the WithClauses; or if these methods are on the children of IndexableLoop such as Forall. We can repeat the interface for getting the WithClauses for something like Cobegin even though it is not a loop.

Yes, we could store the list of WithClauses as its own uAST node, but I don't see the point of this. Forall and Cobegin will still need to have separate methods returning them in the has-a situation; why can't the separate methods handle returning elements of a list instead?

vasslitvinov commented:
> my request that the ForLoop include some sort of Decls for the index variables

I like this idea, too.

> I think it's important that in the common cases along the lines of for (a,b) in zip(A, B) that there are no tuples introduced into the representation.

We strive to preserve the syntax of the source code in uAST, right? The syntax is either an ID or a tuple, rather than "1 or more IDs". If we replace it with a list of variables, we are removing the tuple and obscuring how the user wrote it.

I think having the index thing be represented with a Decl offers a good solution here.

mppf commented:
In discussion with @e-kayrakli and @dlongnecke-cray I think we concluded it's better to have 1 AST slot for the Index Variable(s) Decl and 1 AST slot for the iterand(s); where the index variables might be a TupleDecl and the iterands might be a Zip. @vasslitvinov it sounds like that is the direction you are leaning as well.

The correspondence between index variables and iterands in simple cases like `for (a,b) in zip(A, B) ` should still be relatively easy to establish in any analysis working on the uAST.

It is important in this idea that during lowering from uAST we don't introduce tuples when we don't have to.

dlongnecke-cray commented:
Hi I have some followup questions.

Let's say that in the current scheme there will be 1 index variable node and 1 iterand node.

For the `IterableLoop` interface, it makes sense how to get the index variable declaration. You just do...

```cpp
  /**
    Returns the index variable declaration of this indexable loop.
  */
  const Decl* indexVariableDecl() const {
    if (indexVariableDeclChildNum_ < 0) return nullptr;
    return (const Decl*) children_[indexVariableDeclChildNum_];
  }
```

Michael (@mppf) proposed a way to get at the variable itself...

```
  /**
    Returns the index variable of this indexable loop.
  */
  const Variable* indexVariable() const {
  ```

But what about when the thing that is declared is a `TupleDecl`? Should I expose a way to get the `i`th variable as a convenience?

mppf commented:
> Michael (@mppf) proposed a way to get at the variable itself...
> But what about when the thing that is declared is a TupleDecl? Should I expose a way to get the ith variable as a convenience?

Let's leave this out for now.

dlongnecke-cray commented:
So only provide a way to get the index variable `Decl` but leave it at that for now?

e-kayrakli commented:
I believe I mentioned this in chat, but wanted to note it here:

> But what about when the thing that is declared is a TupleDecl? Should I expose a way to get the ith variable as a convenience?

I would like to have this interface.

> Let's leave this out for now.

Since I am not using the new compiler at the moment, I am OK if it doesn't happen right now :)

dlongnecke-cray commented:
I am certain that we'll add such an interface later down the road. But for now, probably best to only add the bare essentials until things start to come together more.
